pydrive2/fs/spec.py

```diff
@@ -0,0 +1,477 @@
+
+        if detail:
+            return contents
+        else:
+            return [content["name"] for content in contents]
+
+    def find(self, path, detail=False, **kwargs):
+        bucket, base = self.split_path(path)
+
+        seen_paths = set()
+        dir_ids = [self._ids_cache["ids"].copy()]
+        contents = []
+        while dir_ids:
+            query_ids = {
+                dir_id: dir_name
+                for dir_id, dir_name in dir_ids.pop().items()
+                if posixpath.commonpath([base, dir_name]) == base
+                if dir_id not in seen_paths
+            }
+            if not query_ids:
+                continue
+
+            seen_paths |= query_ids.keys()
+
+            new_query_ids = {}
+            dir_ids.append(new_query_ids)
+            for item in self._gdrive_list_ids(query_ids):
+                parent_id = item["parents"][0]["id"]
+                item_path = posixpath.join(query_ids[parent_id], item["title"])
+                if item["mimeType"] == FOLDER_MIME_TYPE:
+                    new_query_ids[item["id"]] = item_path
+                    self._cache_path_id(item_path, item["id"])
+                    continue
+
+                contents.append(
+                    {
+                        "name": posixpath.join(bucket, item_path),
+                        "type": "file",
+                        "size": int(item["fileSize"]),
+                        "checksum": item["md5Checksum"],
+                    }
+                )
+
+        if detail:
+            return contents
+        else:
+            return [content["name"] for content in contents]
+
+    def upload_fobj(self, stream, rpath, callback=None, **kwargs):
+        parent_id = self._get_item_id(self._parent(rpath), create=True)
+        if callback:
+            stream = CallbackIOWrapper(
+                callback.relative_update, stream, "read"
+            )
+        return self.gdrive_upload_fobj(
+            posixpath.basename(rpath.rstrip("/")), parent_id, stream
+        )
+
+    def put_file(self, lpath, rpath, callback=None, **kwargs):
+        if callback:
+            callback.set_size(os.path.getsize(lpath))
+        with open(lpath, "rb") as stream:
+            self.upload_fobj(stream, rpath, callback=callback)
+
+    @_gdrive_retry
+    def gdrive_upload_fobj(self, title, parent_id, stream, callback=None):
+        item = self.client.CreateFile(
+            {"title": title, "parents": [{"id": parent_id}]}
+        )
+        item.content = stream
+        item.Upload()
+        return item
+
+    def cp_file(self, lpath, rpath, **kwargs):
+        """In-memory streamed copy"""
+        with self.open(lpath) as stream:
+            # IterStream objects doesn't support full-length
+            # seek() calls, so we have to wrap the data with
+            # an external buffer.
+            buffer = io.BytesIO(stream.read())
+            self.upload_fobj(buffer, rpath)
```

Comment:
I don't get this. This will fail if a file is bigger than the system RAM.

```suggestion
        with self.open(lpath) as stream:
            self.upload_fobj(stream, rpath)
```

and fix `self.open` to return a proper stream.